source/common/upstream/conn_pool_map_impl.h

```diff
@@ -74,20 +88,42 @@ template <typename KEY_TYPE, typename POOL_TYPE> void ConnPoolMap<KEY_TYPE, POOL
 }
 
 template <typename KEY_TYPE, typename POOL_TYPE>
-void ConnPoolMap<KEY_TYPE, POOL_TYPE>::addDrainedCallback(const DrainedCb& cb) {
+void ConnPoolMap<KEY_TYPE, POOL_TYPE>::addIdleCallback(const IdleCb& cb) {
   Common::AutoDebugRecursionChecker assert_not_in(recursion_checker_);
   for (auto& pool_pair : active_pools_) {
-    pool_pair.second->addDrainedCallback(cb);
+    pool_pair.second->addIdleCallback(cb);
   }
 
   cached_callbacks_.emplace_back(std::move(cb));
 }
 
+template <typename KEY_TYPE, typename POOL_TYPE>
+void ConnPoolMap<KEY_TYPE, POOL_TYPE>::startDrain() {
+  // Copy the `active_pools_` so that it is safe for the call to result
+  // in deletion, and avoid iteration through a mutating container.
```

Comment:
didn't you guard against deletion? is this left over?

Reply:
startDrain() causes idle connections to be immediately closed. If the conn pool is then idle, the idleCallback is invoked under this callstack, which results in the conn pool being removed from this container. So this code is needed.